// term-schemes/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TermSchemes;
(function (TermSchemes) {
    TermSchemes["hyper"] = "hyper";
    TermSchemes["iterm2"] = "iterm2";
    TermSchemes["konsole"] = "konsole";
    TermSchemes["remmina"] = "remmina";
    TermSchemes["terminal"] = "terminal";
    TermSchemes["terminator"] = "terminator";
    TermSchemes["termite"] = "termite";
    TermSchemes["tilda"] = "tilda";
    TermSchemes["xcfe"] = "xcfe";
    TermSchemes["xresources"] = "xresources";
    TermSchemes["xterm"] = "xterm";
})(TermSchemes = exports.TermSchemes || (exports.TermSchemes = {}));
;
var hyper_1 = require("./parsers/hyper");
exports.hyper = hyper_1.hyper;
var iterm2_1 = require("./parsers/iterm2");
exports.iterm2 = iterm2_1.iterm2;
var konsole_1 = require("./parsers/konsole");
exports.konsole = konsole_1.konsole;
var remmina_1 = require("./parsers/remmina");
exports.remmina = remmina_1.remmina;
var terminal_1 = require("./parsers/terminal");
exports.terminal = terminal_1.terminal;
var terminator_1 = require("./parsers/terminator");
exports.terminator = terminator_1.terminator;
var termite_1 = require("./parsers/termite");
exports.termite = termite_1.termite;
var tilda_1 = require("./parsers/tilda");
exports.tilda = tilda_1.tilda;
var xfce_1 = require("./parsers/xfce");
exports.xfce = xfce_1.xfce;
var xresources_1 = require("./parsers/xresources");
exports.xresources = xresources_1.xresources;
var xterm_1 = require("./parsers/xterm");
exports.xterm = xterm_1.xterm;
